### ISSUE TITLE
Cluster-wide alarm state is not updated when alarmed node goes down

### ISSUE THREAD

**Issue by michaelklishin** [rabbitmq/rabbitmq-server#362]
See [this thread](https://groups.google.com/forum/#!topic/rabbitmq-users/py4o6Ia9VLk).

**Comment by binarin** [comment]
I think that alarm list is actually being updated in this case. The problem is that in case of alarmed node brutal death, blocked publishers don't receive 'connection.unblocked' frame. Publishers connected after the moment of death don't have any problems.

So the problem is that 'connection.unblocked' is received only when alarm is explicitly cleared on live node.

**Comment by michaelklishin** [comment]
@binarin also a possibility.

**Comment by carlhoerberg** [comment]
According to my observations connections were still marked as "blocked" in the connections mgmt view, if that gives a clue or rules something out..

**Comment by binarin** [comment]
There are 2 problems there:

1. on node_down rabbit_alarm notifies its subscribers with single notification with Source=[]. If instead notification will be sent for every type of alarm from down node, it will solve the case from original message.
2. rabbit_alarm doesn't send information to its subscribers about which node the alarm corresponds to. As a result I was able to find another bug: set disk alarm for 2 of 3 nodes, try to pubsish message and become blocked, and after clearing disk alarm only on one node it becames unblocked. But I think in this case the publisher should be unlocked until alarms an all nodes were cleared. Fixing this will require more work: rabbit_alarm subscription protocol should be updated with node information, and node information should be saved along with alarm type in rabbit_reader `#throttle{}`.

**Comment by noahhaon** [comment]
I have also observed case 2 mentioned by @binarin above. :+1:

**Comment by michaelklishin** [comment]
Please file a separate issue for case two.

**Comment by noahhaon** [comment]
@michaelklishin done - #379

**Comment by michaelklishin** [comment]
@binarin would you be interested in looking into a PR for case 1? We'll handle case 2 because it involves protocol method changes. Both are only reasonable for `3.6.0`.

**Comment by michaelklishin** [comment]
OK, so maybe I've misread case 2. If it doesn't require a client-server protocol change, even better. However, it does change the internal protocol so can break earlier `3.5.x` releases in mixed clusters => still can only go into `3.6.0`.

**Comment by binarin** [comment]
The more I look into rabbit_alarm, the messier it appears )

Actually, I think I want to solve both problems with one patch, because simple fix for the first issue will not be final.

I'll try to make a draft patch tommorow, so there will be something to discuss.

**Comment by michaelklishin** [comment]
A single PR that fixes both is acceptable. Thank you.

Note that I'm away next week for OpenStack Summit Tokyo, so PRs submitted late this and next week may see little activity for a while.

**Comment by binarin** [comment]
Currently I'm thinking about splitting rabbit_alarm into 2 parts: that handles local-only alarms (`file_descriptor_limit`) and another one that handles cluster-wide alarms (`resource_limit`). The main reason is that currently the `alarmed_nodes` and `alarms` are sometimes not consitent inside `#alarms{}`, and the effort to make them consistent is not worth it (in my opinion). Is this OK?

**Comment by michaelklishin** [comment]
@binarin sounds reasonable.

**Comment by michaelklishin** [comment]
I believe this is addressed by #395 (née #381) by @binarin. Thank you!
